_pydevd_sys_monitoring/_pydevd_sys_monitoring.py

```diff
@@ -168,51 +169,67 @@ def _get_bootstrap_frame(depth: int) -> Tuple[Optional[FrameType], bool]:
 
         return f_bootstrap, is_bootstrap_frame_internal
 
-
 # fmt: off
 # IFDEF CYTHON
-# cdef _get_unhandled_exception_frame(int depth):
+# cdef _is_user_frame(frame: FrameType):
 # ELSE
-def _get_unhandled_exception_frame(depth: int) -> Optional[FrameType]:
+def _is_user_frame(frame: FrameType) -> bool:
 # ENDIF
 # fmt: on
-    try:
-        return _thread_local_info.f_unhandled
-    except:
-        frame = _getframe(depth)
-        f_unhandled = frame
+    if frame is None:
+        return False
 
-        while f_unhandled is not None and f_unhandled.f_back is not None:
-            f_back = f_unhandled.f_back
-            filename = f_back.f_code.co_filename
-            name = splitext(basename(filename))[0]
+    filename = frame.f_code.co_filename
+    name = splitext(basename(filename))[0]
 
-            # When the back frame is the bootstrap (or if we have no back
-            # frame) then use this frame as the one to track.
-            if name == "threading":
-                if f_back.f_code.co_name in ("__bootstrap", "_bootstrap", "__bootstrap_inner", "_bootstrap_inner", "run"):
-                    break
+    # When the frame is the bootstrap it is not a user frame.
+    if name == "threading":
+        if frame.f_code.co_name in ("__bootstrap", "_bootstrap", "__bootstrap_inner", "_bootstrap_inner", "run"):
+            return False
 
-            elif name == "pydev_monkey":
-                if f_back.f_code.co_name == "__call__":
-                    break
+    elif name == "pydev_monkey":
+        if frame.f_code.co_name == "__call__":
+            return False
 
-            elif name == "pydevd":
-                if f_back.f_code.co_name in ("_exec", "run", "main"):
-                    break
+    elif name == "pydevd":
+        if frame.f_code.co_name in ("_exec", "run", "main"):
+            return False
 
-            elif name == "pydevd_runpy":
-                if f_back.f_code.co_name.startswith(("run", "_run")):
-                    break
+    elif name == "pydevd_runpy":
+        if frame.f_code.co_name.startswith(("run", "_run")):
+            return False
 
-            f_unhandled = f_back
+    elif filename == "<frozen runpy>":
+        if frame.f_code.co_name.startswith(("run", "_run")):
+            return False
 
-        if f_unhandled is not None:
-            _thread_local_info.f_unhandled = f_unhandled
-            return _thread_local_info.f_unhandled
+    elif name == 'runpy':
+        if frame.f_code.co_name.startswith(("run", "_run")):
+            return False
 
-        return f_unhandled
+    return True
 
+# fmt: off
+# IFDEF CYTHON
+# cdef _is_last_user_frame(frame: FrameType):
+# ELSE
+def _is_last_user_frame(frame: FrameType) -> bool:
```

Comment:
I changed the logic for get_unhandled_frame to just check if the passed in frame was the last user frame. When using Cython the old logic wasn't working because getFrame wasn't returning any of the pydevd frames and the logic was off.

Reply:
The tests that were failing were these:

https://github.com/microsoft/debugpy/blob/25955a05d87a6e3d7694f7479c1c17ded0f9af90/tests/debugpy/test_exception.py#L160

The unhandled frame logic (when Cython was active) would fail to catch that something was unhandled.

Reply:
Sorry still need to do this too. Haven't gotten to debugging this with the old code yet.

Reply:
Going back to the old way, all of these tests fail in debugpy:

```
FAILED tests/debugpy/test_exception.py::test_systemexit[0-zero-uncaught-raised-attach_connect(cli)-program]
... (38 in total)
```

I'm looking into how to create a similar test in PyDev.Debugger.

Reply:
Hmm, there's already a bunch of `test_case_sys_exit_unhandled_exception` type cases. Not sure why they have different results.

Reply:
I added some tests that simulate what happens in debugpy failures but they don't fail with the old code. It seems the old code only fails on 'attach'. (which I haven't figured out how to do in pydevd yet)

Would attach change how sys.getframe worked?

Reply:
Okay figured out how to write an attach test. Still working on how this is failing with the old code

Reply:
Still doesn't fail with pydevd. Blech. I'll have to debug the difference in debugpy then.

Reply:
Okay in debugpy these attach tests throw more unhandled exceptions, maybe because jmc isn't being applied?

I can see a whole bunch more exceptions:

```
0.00s - Ignore exception <class 'AttributeError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (__str__)
0.00s - Ignore exception <class 'AttributeError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (drive)
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (open)
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (open)
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (read_text)
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\pathlib.py -- (read_text)
0.00s - File not traced (not in project): C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\site-packages\importlib_metadata\__init__.py
0.00s - Ignore exception <class 'FileNotFoundError'> in library C:\Users\rchiodo\AppData\Local\anaconda3\envs\py312_64\Lib\site-packages\importlib_metadata\__init__.py -- (read_text)
... // There's a bunch more
```

This results in this code here getting hit earlier in `_get_unhandled_exception_frame`

```python

        if f_unhandled is not None:
            _thread_local_info.f_unhandled = f_unhandled
            return _thread_local_info.f_unhandled
```

But there's nothing that ever unsets this `f_unhandled` for the current thread. So it stays there. 

This prevents future exceptions from being marked as unhandled because this exception frame is reused. 

Removing this caching fixes the problem.

Trying to figure out a way to reproduce this in a test. Maybe have the test throw more exceptions in a different frame before the sys.exit.

Reply:
I think that the problem may be that it's just caching one when possibly other entries arise?

The logic here is that the cache is caching the first frame where things should be considered unhandled

So, in a stack such as

```
runpy.run
user code
```
it should mark `runpy.run` as the top thing, but there are cases when multiple top frames could be considered

i.e.:

```
threading.bootstrap1
threading.bootstrap2
user code
```

in this case all of those could be considered

So, the logic could be changed to keep on going up the stack and instead of keeping just 1 keep all that match to see if it works (so, instead of `_get_unhandled_exception_frame` it'd be `_get_unhandled_exception_frames`).

If that doesn't work, please just comment the cache out instead of doing the refactor and then this can be checked after the PR is merged.

Reply:
Why is the cache needed in the first place? It seems like everytime an unwind event occurred it should be reset as it could be for an entirely different exception. At the very least, shouldn't the cache also keep track of the owning exception?

Reply:
I updated the cache to clear itself it the cached frame isn't in the callstack for the new frame (which I believe should mean that the old frame was from a different exception).

Reply:
Well, the cache is a performance improvement thing (so, for correctness we could remove it for now, but for performance it may be good to add it back in the future if possible).

A note: in the refactoring you called it `_is_user_frame`, but that's not really correct, what it's trying to discover here is handling for unhandled exceptions and it's trying to determine if this is the frame that when reached the exception has to be considered unhandled (meaning it's the last frame before actually being raised as unhandled) -- the thing is that (in theory) this shouldn't really change for the same thread after it's computed once (apart from the corner case you found where apparently this is being computed too soon).